private/Metaprogram.h

```diff
@@ -0,0 +1,349 @@
+   using PankuDependencies = PankuProcessedList::headless_type; \
+   using PankuPath = PankuMetaprogram::topological_sort<PankuUserClassList, PankuDependencies>::type; \
+   using PankuPathPointerised = PankuMetaprogram::type_list_pointerise<PankuPath>::type; \
+   using PankuClassTuple = convert<PankuPathPointerised, std::tuple>; \
+   extern Panku<PankuClassTuple> panku;
+
+#define PANKU_INSTANCE Panku<PankuClassTuple> panku;
+
+#define STANDALONE(ClassName) type_list<ClassName>
+#define DEPENDENCY(ClassName, ...) type_list<ClassName, ##__VA_ARGS__>
+
+using namespace TypeList;
+
+template<typename UserClass>
+UserClass& ConstructAndInitialise();
+
+template<typename Tuple>
+class Panku
+{
+   public:
+   Panku() 
+   {
+      TupleManipulation::for_each_in_tuple(userClassTuple, [](auto& element)
+            { element = &ConstructAndInitialise<decltype(*element)>(); });
+   }
+
+   template<typename UserClass>
+   UserClass& Get() 
+   {
+      auto userObjectPointer = std::get<UserClass*>(userClassTuple);
+      assert(userObjectPointer);
+      return *userObjectPointer;
+   }
+
+   private:
+   Tuple userClassTuple;
+};
+
+namespace PankuMetaprogram
+{
+
+   // Turns a type list into a type list of its pointer types.
+   template<class List>
+   struct type_list_pointerise;
+
+   template<>
+   struct type_list_pointerise<type_list<>>
+   {
+      using type = type_list<>;
+   };
+
+   template<typename T, typename... Ts>
+   struct type_list_pointerise<type_list<T, Ts...>>
+   {
+      using type = typename type_list_push_front<T*, typename type_list_pointerise<type_list<Ts...>>::type>::type;
+   };
+
+   // Extracts the first type from each list in a list of lists
+   template <class ListOfLists>
+   struct type_list_extract_heads;
+
+   template<>
+   struct type_list_extract_heads<type_list<>>
+   {
+      using type = type_list<>;
+      using headless_type = type_list<>;
+   };
+
+   template<typename T, typename... OtherLists>
+   struct type_list_extract_heads<type_list<type_list<T>, OtherLists...>>
+   {
+      using type = typename type_list_push_front<T, typename type_list_extract_heads<type_list<OtherLists...>>::type>::type;
+      using headless_type = typename type_list_push_front<type_list<>, typename type_list_extract_heads<type_list<OtherLists...>>::headless_type>::type;
+   };
+
+   template<typename T, typename... Ts, typename... OtherLists>
+   struct type_list_extract_heads<type_list<type_list<T, Ts...>, OtherLists...>>
+   {
+      using type = typename type_list_push_front<T, typename type_list_extract_heads<type_list<OtherLists...>>::type>::type;
+      using headless_type = typename type_list_push_front<type_list<Ts...>, typename type_list_extract_heads<type_list<OtherLists...>>::headless_type>::type;
+   };
+
+   ////////////////////////////////////////////////
+   // Check all elements of list A are in list B //
+   ////////////////////////////////////////////////
```

Comment:
Yeah... I should namespace it I suppose, I want the comment to apply to a few meta-structs.